src/realm/object-store/sync/app.cpp

```diff
@@ -723,85 +744,180 @@ std::string App::url_for_path(const std::string& path = "") const
+        return completion(util::none); // early return
+    }
+    else {
+        std::lock_guard<std::mutex> lock(*m_route_mutex);
+        route = util::format("%1/location", new_hostname ? get_app_route(new_hostname) : get_app_route());
+    }
 
     Request req;
     req.method = HttpMethod::get;
     req.url = route;
     req.timeout_ms = m_request_timeout_ms;
 
-    m_config.transport->send_request_to_server(
-        std::move(req), [this, completion = std::move(completion)](const Response& response) {
-            try {
-                auto json = parse<BsonDocument>(response.body);
-                auto hostname = get<std::string>(json, "hostname");
-                auto ws_hostname = get<std::string>(json, "ws_hostname");
-                auto deployment_model = get<std::string>(json, "deployment_model");
-                auto location = get<std::string>(json, "location");
-                m_sync_manager->perform_metadata_update([&](const SyncMetadataManager& manager) {
-                    manager.set_app_metadata(deployment_model, location, hostname, ws_hostname);
-                });
-
-                auto metadata = m_sync_manager->app_metadata();
-
-                std::lock_guard<std::mutex> lock(*m_route_mutex);
-                m_base_route = hostname + base_path;
-                std::string this_app_path = app_path + "/" + m_config.app_id;
-                m_app_route = m_base_route + this_app_path;
-                m_auth_route = m_app_route + auth_path;
-                m_sync_manager->set_sync_route(ws_hostname + base_path + this_app_path + sync_path);
-            }
-            catch (const AppError&) {
-                return completion(std::move(response));
-            }
+    m_config.transport->send_request_to_server(std::move(req), [this, completion = std::move(completion)](
+                                                                   const Request&, const Response& response) {
+        // If the response contains an error, then pass it up
+        if (response.http_status_code >= 300 || (response.http_status_code < 200 && response.http_status_code != 0)) {
+            return completion(std::move(response)); // early return
+        }
 
-            completion(util::none);
-        });
+        try {
+            auto json = parse<BsonDocument>(response.body);
+            auto hostname = get<std::string>(json, "hostname");
+            auto ws_hostname = get<std::string>(json, "ws_hostname");
+            auto deployment_model = get<std::string>(json, "deployment_model");
+            auto location = get<std::string>(json, "location");
+            m_sync_manager->perform_metadata_update([&](SyncMetadataManager& manager) {
+                manager.set_app_metadata(deployment_model, location, hostname, ws_hostname);
+            });
+
+            update_hostname(m_sync_manager->app_metadata());
+        }
+        catch (const AppError&) {
```

Comment:
just curious what kind of AppError's we could catch here?

Reply:
The `parse` and `get` functions throw an `AppError` with a `JSONErrorCode` if the operation fails. Since the only interface to the caller is a callback that takes a Response object, the failure will be determined (again) when the callback is run.